File: Simulator-CSharp/Views/CurrencyBox.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ATM.Views
{
    public partial class CurrencyBox : UserControl
    {
        public CurrencyBox()
        {
            InitializeComponent();
        }

        #region Properties

        private double _Total = 0.0;
        public double Total
        {
            get { return _Total; }
        }

        private double _Value;
        public double Value
        {
            get { return _Value; }
            set
            {
                if (_Value != value)
                {
                    _Value = value;
                    OnValueChanged(EventArgs.Empty);
                }
            }
        }
        private int _Amount;
        public int Amount
        {
            get { return _Amount; }
            set
            {
                if (_Amount != value)
                {
                    _Amount = value;
                    OnAmountChanged(EventArgs.Empty);
                }
            }
        }

        private int _Minimum = 0;
        public int Minimum
        {
            get { return _Minimum; }
            set
            {
                if (_Minimum != value)
                {
                    _Minimum = value;
                    OnMinimumChanged(EventArgs.Empty);
                }
            }
        }
        private int _Maximum = 100;
        public int Maximum
        {
            get { return _Maximum; }
            set
            {
                if (_Maximum != value)
                {
                    _Maximum = value;
                    OnMaximumChanged(EventArgs.Empty);
                }
            }
        }

        #endregion

        #region Events

        private void OnMinimumChanged(EventArgs e)
        {
            amountControl.Minimum = Minimum;
        }

        private void OnMaximumChanged(EventArgs e)
        {
            amountControl.Maximum = Maximum;
        }

        private void OnAmountChanged(EventArgs e)
        {
            amountControl.Value = Amount;
            CalcTotal();
        }

        private void OnValueChanged(EventArgs e)
        {
            comboBoxDenomination.SelectedItem = Value.ToString();
            CalcTotal();
        }

        private void CalcTotal()
        {
            double _result = 0.0;
            if (comboBoxDenomination.SelectedItem != null && !string.IsNullOrEmpty(comboBoxDenomination.SelectedItem.ToString()))
            {
                _result = Convert.ToDouble(comboBoxDenomination.SelectedItem) * amountControl.Value;
            }

            if (_result != this._Total)
            {
                this._Total = _result;
                if (ValueChanged != null)
                {
                    ValueChanged(this, EventArgs.Empty);
                }
            }

        }

        public event EventHandler ValueChanged;

        #endregion

        private void SubTotal()
        {
            textBoxTotal.Text = string.Format("{0:C}", Total);
            double _percent = Percent();

            if (_percent == 0)
            {
                BackColor = SystemColors.Control;
            }
            else if (_percent > 0 && _percent < 25)
            {
                BackColor = Color.FromArgb(245, 68, 55);
            }
            else if (_percent >= 25 && _percent < 50)
            {
                BackColor = Color.FromArgb(255, 153, 0);
            }
            else {
                BackColor = Color.FromArgb(77, 176, 81);
            }
        }

        public double Percent()
        {
            return amountControl.Percent();
        }

        private void amountControl_ValueChanged(object sender, EventArgs e)
        {
            Amount = amountControl.Value;
            SubTotal();
        }

        private void comboBoxDenomination_SelectedIndexChanged(object sender, EventArgs e)
        {
            Value = Double.Parse(comboBoxDenomination.SelectedItem.ToString());
            SubTotal();
        }
    }
}

File: Simulator-VB/Views/CurrencyBox.vb
Public Class CurrencyBox

#Region "Properties"

    Private _Total As Double = 0.0
    Public ReadOnly Property Total As Double
        Get
            Return _Total
        End Get
    End Property

    Private _Value As Double
    Public Property Value As Double
        Get
            Return _Value
        End Get
        Set(value As Double)
            If _Value <> value Then
                _Value = value
                OnValueChanged(EventArgs.Empty)
            End If
        End Set
    End Property
    Private _Amount As Integer
    Public Property Amount As Integer
        Get
            Return _Amount
        End Get
        Set(value As Integer)
            If _Amount <> value Then
                _Amount = value
                OnAmountChanged(EventArgs.Empty)
            End If
        End Set
    End Property

    Private _Minimum As Integer = 0
    Public Property Minimum As Integer
        Get
            Return _Minimum
        End Get
        Set(value As Integer)
            If _Minimum <> value Then
                _Minimum = value
                OnMinimumChanged(EventArgs.Empty)
            End If
        End Set
    End Property
    Private _Maximum As Integer = 100
    Public Property Maximum As Integer
        Get
            Return _Maximum
        End Get
        Set(value As Integer)
            If _Maximum <> value Then
                _Maximum = value
                OnMaximumChanged(EventArgs.Empty)
            End If
        End Set
    End Property

#End Region

    Sub New()

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().
        'Me.ComboBoxDenomination.DataBindings.Add("SelectedItem", Me, "Value", False, DataSourceUpdateMode.OnPropertyChanged, Nothing, "{0}")
        'Me.AmountControl.DataBindings.Add("Value", Me, "Amount", False, DataSourceUpdateMode.OnPropertyChanged)

    End Sub

#Region "Events"

    Private Sub OnMinimumChanged(e As EventArgs)
        Me.AmountControl.Minimum = Me.Minimum
    End Sub

    Private Sub OnMaximumChanged(e As EventArgs)
        Me.AmountControl.Maximum = Me.Maximum
    End Sub

    Private Sub OnAmountChanged(e As EventArgs)
        Me.AmountControl.Value = Me.Amount
        CalcTotal()
    End Sub

    Private Sub OnValueChanged(e As EventArgs)
        Me.ComboBoxDenomination.SelectedItem = Me.Value.ToString()
        CalcTotal()
    End Sub

    Private Sub CalcTotal()
        Dim _result As Double = 0.0
        If Not String.IsNullOrEmpty(Me.ComboBoxDenomination.SelectedItem) Then
            _result = CDbl(Me.ComboBoxDenomination.SelectedItem) * Me.AmountControl.Value
        End If

        If _result <> Me._Total Then
            Me._Total = _result
            RaiseEvent ValueChanged(Me, EventArgs.Empty)
        End If

    End Sub

    Public Event ValueChanged As EventHandler

#End Region

    Private Sub SubTotal()
        Me.TextBoxTotal.Text = String.Format("{0:C}", Me.Total)
        Dim _percent As Double = Me.Percent

        If _percent = 0 Then
            Me.BackColor = SystemColors.Control
        ElseIf _percent > 0 AndAlso _percent < 25 Then
            Me.BackColor = Color.FromArgb(245, 68, 55)
        ElseIf _percent >= 25 AndAlso _percent < 50
            Me.BackColor = Color.FromArgb(255, 153, 0)
        Else
            Me.BackColor = Color.FromArgb(77, 176, 81)
        End If
    End Sub

    Public Function Percent() As Double
        Return Me.AmountControl.Percent()
    End Function

    Private Sub AmountControl_ValueChanged(sender As Object, e As EventArgs) Handles AmountControl.ValueChanged
        Me.Amount = Me.AmountControl.Value
        SubTotal()
    End Sub

    Private Sub ComboBoxDenomination_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ComboBoxDenomination.SelectedIndexChanged
        Me.Value = CDbl(Me.ComboBoxDenomination.SelectedItem)
        SubTotal()
    End Sub


End Class
